File: _site/python_solutions/5087.letter-tile-possibilities.js
var bisect_left = function(nums, target) {
    var i = 0,
        j = nums.length - 1,
        mid;
    while (i < j) {
        mid = ~~((i + j) / 2);
        if (nums[mid] < target)
            i = mid + 1;
        else
            j = mid;
    }
    return i;
}

var bisect_right = function(nums, target) {
    var i = 0,
        j = nums.length - 1,
        mid;
    while (i < j) {
        mid = Math.ceil((i + j) / 2);
        if (nums[mid] > target)
            j = mid - 1;
        else
            i = mid;
    }
    return j;
}

var two_d_array = function(m, n) {
    return new Array(m).fill(0).map(() => new Array(n).fill(0));
}

var reverse = function(s) {
    return [...str].reverse().join('');
}


// simple print
var print = function(a) {
    console.log(a);
}

// For counting elements in a list 
var counter = function(array_or_string) {
    var dict = {};
    for (let i = 0; i < array_or_string.length; i++) {
        let e = array_or_string[i];
        dict[e] = dict[e] ? dict[e] += 1 : 1;
    }
    return dict;
}

var sum = function(array) {
    return array.reduce(function(a, b) {
        return a + b;
    }, 0);
}

var isodd = function(n) {
    return n % 2 == 1;
}

var iseven = function(n) {
    return n % 2 == 0;
}

var list = function(size, value) {
    return Array(size).fill(value);
}


var len = function(str_or_array) {
    return str_or_array.length;
}

var sorted = function(arr) {
    return arr.sort((a, b) => (a - b));
}

// print out a Map 
function printMap(m) {
    m.forEach(function(v, k) {
        console.log(k + " " + v);
    })
}

var gcd = function(a, b) {
    return b == 0 ? a : gcd(b, a % b);
}

var say = function(a) {
    console.log(a)
};


var two_d_array = function(m, n, v) {
    return [...Array(m)].map(x => Array(n).fill(v));
}

function floor(float_number) {
    return Math.floor(float_number)
}

function ceil(float_number) {
    return Math.ceil(float_number)
}


var max = function(arr) {
    return Math.max(...arr);
}

var min = function(arr) {
    return Math.min(...arr)
};

var pairmax = function(a, b) {
    return Math.max(a, b)
}

var pairmin = function(a, b) {
    return Math.min(a, b)
}

// create an array with elements from 0 to n
var nlist = function(n) {
    return [...Array(n + 1).keys()];
}


/*
 * @lc app=leetcode id=5087 lang=javascript
 *
 * [5087] Letter Tile Possibilities
 *
 * https://leetcode.com/problems/letter-tile-possibilities/description/
 *
 * algorithms
 * Medium (76.77%)
 * Total Accepted:    2K
 * Total Submissions: 2.6K
 * Testcase Example:  '"AAB"'
 *
 * You have a set of tiles, where each tile has one letter tiles[i] printed on
 * it.  Return the number of possible non-empty sequences of letters you can
 * make.
 * 
 * 
 * 
 * Example 1:
 * 
 * 
 * Input: "AAB"
 * Output: 8
 * Explanation: The possible sequences are "A", "B", "AA", "AB", "BA", "AAB",
 * "ABA", "BAA".
 * 
 * 
 * 
 * Example 2:
 * 
 * 
 * Input: "AAABBC"
 * Output: 188
 * 
 * 
 * 
 * 
 * 
 * Note:
 * 
 * 
 * 1 <= tiles.length <= 7
 * tiles consists of uppercase English letters.
 * 
 * 
 */
/**
 * @param {string} tiles
 * @return {number}
 */
var numTilePossibilities = function(tiles) {
    var c = list(26, 0);
    for (let i = 0; i < len(tiles); i++) c[tiles[i].charCodeAt(0) - 65] += 1;

    var dfs = function(cter) {
        // print(cter)
        var mus = 0;
        for (let i = 0; i < 26; i++) {
            if (cter[i] > 0) {
                mus += 1;
                cter[i] -= 1;
                mus += dfs(cter);
                cter[i] += 1;
            }

        }
        return mus;
    }

    return dfs(c);
};

print(numTilePossibilities('AAABBC'))

File: _site/python_solutions/5087.letter-tile-possibilities.rb
#
# @lc app=leetcode id=5087 lang=ruby
#
# [5087] Letter Tile Possibilities
#
# https://leetcode.com/problems/letter-tile-possibilities/description/
#
# algorithms
# Medium (76.77%)
# Total Accepted:    2K
# Total Submissions: 2.6K
# Testcase Example:  '"AAB"'
#
# You have a set of tiles, where each tile has one letter tiles[i] printed on
# it.  Return the number of possible non-empty sequences of letters you can
# make.
#
#
#
# Example 1:
#
#
# Input: "AAB"
# Output: 8
# Explanation: The possible sequences are "A", "B", "AA", "AB", "BA", "AAB",
# "ABA", "BAA".
#
#
#
# Example 2:
#
#
# Input: "AAABBC"
# Output: 188
#
#
#
#
#
# Note:
#
#
# 1 <= tiles.length <= 7
# tiles consists of uppercase English letters.
#
#
#
# @param {String} tiles
# @return {Integer}

# t is either a string or an array
def counter(t)
  cter = Hash.new(0)
  arr = t.is_a?(String) ? t.chars : t
  arr.each do |c|
    cter[c] += 1
  end
  cter
end

def num_tile_possibilities(tiles)
  c = counter(tiles)
  res = 0
  dfs = lambda do |cter|
    sum = 0
    'A'.upto('Z').each do |ic|
      next unless cter[ic] > 0

      sum += 1
      cter[ic] -= 1
      sum += dfs.call(cter)
      cter[ic] += 1
    end
    sum
  end
  dfs.call(c)
end

tiles = 'AAABBC'
p num_tile_possibilities(tiles)
